includes/sanitizers/class-amp-meta-sanitizer.php

```diff
@@ -0,0 +1,333 @@
+<?php
+/**
+ * Class AMP_Links_Sanitizer.
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Meta_Sanitizer.
+ *
+ * Sanitizes meta tags found in the header.
+ *
+ * @since 1.5.0
+ */
+class AMP_Meta_Sanitizer extends AMP_Base_Sanitizer {
+
+	/**
+	 * Tag.
+	 *
+	 * @var string HTML <meta> tag to identify and replace with AMP version.
+	 */
+	public static $tag = 'meta';
+
+	/**
+	 * Placeholder for default arguments, to be set in child classes.
+	 *
+	 * @var array
+	 */
+	protected $DEFAULT_ARGS = [ // phpcs:ignore WordPress.NamingConventions.ValidVariableName.PropertyNotSnakeCase
+		'use_document_element' => true, // We want to work on the header, so we need the entire document.
+	];
+
+	/**
+	 * Reference to the shared XPath object to query the DOM.
+	 *
+	 * @var DOMXPath
+	 */
+	protected $xpath;
+
+	/**
+	 * The document's <head> element.
+	 *
+	 * @var DOMElement
+	 */
+	protected $head;
+
+	/*
+	 * Tags array keys.
+	 */
+	const TAG_CHARSET        = 'charset';
+	const TAG_VIEWPORT       = 'viewport';
+	const TAG_AMP_SCRIPT_SRC = 'amp_script_src';
+	const TAG_OTHER          = 'other';
+
+	/**
+	 * Associative array of DOMElement arrays.
+	 *
+	 * Each key in the root level defines one group of meta tags to process.
+	 *
+	 * @var array $tags {
+	 *     An array of meta tag groupings.
+	 *
+	 *     @type DOMElement[] $charset        Charset meta tag(s).
+	 *     @type DOMElement[] $viewport       Viewport meta tag(s).
+	 *     @type DOMElement[] $amp_script_src <amp-script> source meta tags.
+	 *     @type DOMElement[] $other          Remaining meta tags.
+	 * }
+	 */
+	protected $meta_tags = [
+		self::TAG_CHARSET        => [],
+		self::TAG_VIEWPORT       => [],
+		self::TAG_AMP_SCRIPT_SRC => [],
+		self::TAG_OTHER          => [],
+	];
+
+	/**
+	 * Charset to use for AMP markup.
+	 *
+	 * @var string
+	 */
+	const AMP_CHARSET = 'utf-8';
+
+	/**
+	 * Viewport settings to use for AMP markup.
+	 *
+	 * @var string
+	 */
+	const AMP_VIEWPORT = 'width=device-width';
+
+	/**
+	 * Sanitize.
+	 */
+	public function sanitize() {
+		$this->xpath = new DOMXPath( $this->dom );
+		$this->head  = $this->ensure_head_is_present();
+		$charset     = $this->detect_charset();
+		$elements    = $this->dom->getElementsByTagName( static::$tag );
+
+		// Remove all nodes for easy reordering later on.
+		$elements = array_map(
+			static function ( $element ) {
+				return $element->parentNode->removeChild( $element );
+			},
+			iterator_to_array( $elements, false )
```

Comment:
TIL! We could be using this lots of places where we're currently iterating over a `DOMNodeList` to push onto an array which we then iterate over to potentially remove elements. Removing elements from the DOM while iterating over a `DOMNodeList` is problematic since it is a live node list.

Reply:
Yes, the `DOMNodeList` does not seem to have a built-in way to turn itself into an array, but any iterator can be turned into one via `iterator_to_array()`.

Array/Traversable/Iterable handling in PHP is just a big mess, unfortunately.